staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go

```diff
@@ -87,13 +89,13 @@ func (t *azureFileCSITranslator) TranslateInTreeInlineVolumeToCSI(volume *v1.Vol
 			Spec: v1.PersistentVolumeSpec{
 				PersistentVolumeSource: v1.PersistentVolumeSource{
 					CSI: &v1.CSIPersistentVolumeSource{
-						Driver:           AzureFileDriverName,
-						VolumeHandle:     fmt.Sprintf(volumeIDTemplate, "", accountName, azureSource.ShareName, ""),
-						ReadOnly:         azureSource.ReadOnly,
-						VolumeAttributes: map[string]string{azureFileShareName: azureSource.ShareName},
-						NodeStageSecretRef: &v1.SecretReference{
-							Name:      azureSource.SecretName,
-							Namespace: defaultSecretNamespace,
+						Driver:       AzureFileDriverName,
+						VolumeHandle: fmt.Sprintf(volumeIDTemplate, "", accountName, azureSource.ShareName, ""),
+						ReadOnly:     azureSource.ReadOnly,
+						VolumeAttributes: map[string]string{
+							azureFileShareName:   azureSource.ShareName,
+							secretNameField:      azureSource.SecretName,
```

Comment:
This means that the CSI driver on the node needs permissions to read secrets (and it can read any secrets)? 

For better security posture, we have the CSI NodeStageSecretRefs so that kubelet and node authorizer can control access to secrets for node entities. So I think it's still better that we figure out how to get node authorizer to support migration.

cc @kubernetes/sig-auth-pr-reviews

Reply:
Azure File CSI driver already supports this(has read secrets permission already), so shall we merge this PR first?

Reply:
This PR just does the in-tree PV translation, I think we could go on this PR first, and then we can use https://github.com/kubernetes/kubernetes/issues/96508 to track the `NodeStageSecretRef` permission issue.

Reply:
I don't think it's a good security practice to give the csi node driver secret read permission, and from a CSI migration perspective is a regression of security requirements on the node (in-tree plugin didn't need broad secret read permissions).

So I think we should work on fixing the CSI driver first and doing the right thing for csi migration.

Reply:
secret read permission is not required in Azure File CSI driver:
if it cannot read secret, node should have azure identity to access Azure account key through Azure API, and then it would leverage azure identity to read account key directly.

This PR just does the in-tree PV translation, it's up to CSI driver config that whether it has secret read permission or not on node.

Reply:
@msau42 Aure file in-tree plugin also needs secret read permission:
https://github.com/kubernetes/kubernetes/blob/bbde53e30054ff93037ebc57edc8d0176f777297/pkg/volume/azure_file/azure_file.go#L280-L282

Reply:
I'm not familiar with Azure identity. Is it able to restrict secret access based on what pods are scheduled to that node?

The main benefit of node authorizer, is that it limits a node's access to only objects that belong to pods scheduled to that node. It's designed to reduce the scope of damage if the node got compromised so that it cannot access use kubelet's permission to access everything in the cluster.

Reply:
if node does not have secret read permission, it would try to use Azure API to access storage account key other than from k8s secret in Azure File CSI driver. The permission issue in node authorizer is that in CSI migration scenario, CSI driver create a secret to store account key while CSI plugin interface in kubelet is not aware of that, in the CSI plugin interface in kubelet, it even don't know the secret name, that secret name is generated in CSI driver.

Read secret on CSI driver node is not a hard requirement, it could still fall back to use Azure API to access storage account key directly if user does not configure kubeConfig on driver node daemonset, so I don't see there is security issue for this PR since it just does not PV translation.

If user wants more security on Azure File CSI driver, they should not configure kubeConfig on driver node daemonset, by that way, this CSI driver would fall back to use Azure API to access storage account key directly.

Reply:
I'm actually wondering about Azure API security vs K8s node authorizer. The k8s node authorizer is able to restrict secret access to only pods scheduled to that node. Would Azure API be able to similarly restrict access if the node got compromised?

I suspect the issue with node authorizer is that it's looking for a real PV object that has the secret reference specified in it. Can we set the [AzureFilePersistentVolumeSource.Secret](https://github.com/kubernetes/kubernetes/blob/3722cef5e5b73131bad38f57b4c5c292e898fc5b/pkg/apis/core/types.go#L1246) when we create a volume for csi migration?

Reply:
Yes, we could revoke azure identity from VM if that node got compromised. 

There is already CSI --> in-tree PV translation in CSI migration, while it does not work:
https://github.com/kubernetes/kubernetes/blob/a098b5b16c9a7f43bebe34a09f31dfa2f2016531/staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go#L180-L181

I would suggest merge this fix first, and then check whether there is better way to leverage `NodeStageSecretRef`, that issue was opened more than 2 months ago, need to make it work first, thanks.

Reply:
the azure volume plugins and CSI translation should use/populate the secret name/namespace fields in the API objects for anything they want the node to be authorized to fetch:

For pods:

https://github.com/kubernetes/kubernetes/blob/d52fe32cd5dab07ec8b93615ed42f6e621c4884d/pkg/api/v1/pod/util.go#L132-L135

https://github.com/kubernetes/kubernetes/blob/d52fe32cd5dab07ec8b93615ed42f6e621c4884d/pkg/api/v1/pod/util.go#L176-L179

For persistent volumes:

https://github.com/kubernetes/kubernetes/blob/d52fe32cd5dab07ec8b93615ed42f6e621c4884d/pkg/api/v1/persistentvolume/util.go#L39-L49

https://github.com/kubernetes/kubernetes/blob/d52fe32cd5dab07ec8b93615ed42f6e621c4884d/pkg/api/v1/persistentvolume/util.go#L128-L137

Using unstructured volumeattributes to plumb secret name/namespace references is not supported by the node authorizer

Reply:
hi @liggitt, that's what we do in the beginning(without this PR), `secretName` and `secretNamespace` is already set, while node authorizer does not work only in CSI migration scenario, detailed issue: https://github.com/kubernetes/kubernetes/issues/96508, this PR did a workaround, and works in CSI migration.

```yaml
$ k get pv pvc-01bd1d0b-b965-4d37-8331-817a296c8998 -o yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubernetes.io/azure-file-resource-group: andy-1191
    pv.kubernetes.io/provisioned-by: file.csi.azure.com
spec:
  accessModes:
  - ReadWriteMany
  azureFile:
    secretName: azure-storage-account-f97cecd00f4d4433ebd6809-secret
    secretNamespace: null
    shareName: pvc-01bd1d0b-b965-4d37-8331-817a296c8998
  capacity:
    storage: 100Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: pvc-azurefile
    namespace: test
  persistentVolumeReclaimPolicy: Delete
  storageClassName: azurefile-456-kubernetes.io-azure-file-dynamic-sc-4hv88
  volumeMode: Filesystem
status:
  phase: Bound
```

Reply:
The node authorizer only permits access to secrets in the pod's namespace based on volume definitions in the pod. That is the case for both azureFile and csi volume definitions in a pod.

It looks like the azurefile volume plugin [looks in the default namespace](https://github.com/kubernetes/kubernetes/blob/da29bd91336db99b132f1aa5d768f99088eaec5d/pkg/volume/azure_file/azure_file.go#L376-L377) for secrets referenced in a pod's `azureFile` volume. That means secret access was not permitted by the node authorizer even in non-CSI migration cases.

If non-CSI migration cases were working for pods outside the default namespace, something else must have been granting access to the node. The fix for the CSI-migration case would be to grant access the same way, not to stop using the API fields intended to communicate the name/namespace coordinates of secrets to use.